framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,901 @@
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.TypeAnnotationPosition;
+import com.sun.tools.javac.code.Type.ArrayType;
+import com.sun.tools.javac.code.Type.ClassType;
+
+/**
+ * SignatureInferenceScenes represents a set of annotations that are inferred
+ * in a program.
+ * This class stores annotations for fields, method return types, and method
+ * parameters.
+ * <p>
+ * The set of annotations inferred for a certain class is stored in an
+ * {@link annotations.el.AScene}, which {@link #writeScenesToJaif} can write
+ * into a .jaif file.
+ * For example, a class field of a class whose fully-qualified name is
+ * {@code my.package.MyClass} will have its inferred type stored in a Scene,
+ * and later written into a file named {@code my.package.MyClass.jaif}.
+ * <p>
+ * This class populates the initial Scenes by reading existing .jaif files
+ * on the {@link #JAIF_FILES_PATH} directory. Having more information in those
+ * initial .jaif files means that the precision achieved by the signature
+ * inference analysis will be better. {@link #writeScenesToJaif} rewrites
+ * the initial .jaif files, and may create new ones.
+ * <p>
+ * Calling an update* method
+ * ({@link #updateInferredFieldType},
+ * {@link #updateInferredMethodParametersTypes}, or
+ * {@link #updateInferredMethodReturnType})
+ * replaces the currently-stored type for an element in a Scene, if any,
+ * by the LUB of it and the update method's argument.
+ * <p>
+ * This class does not store annotations for an element if the element has
+ * explicit annotations:  an update* method ignores an
+ * explicitly annotated field, method return, or method parameter when
+ * passed as an argument.
+ * <p>
+ *  @author pbsf
+ */
+//  TODO: We could add an option to update the type of explicitly annotated
+//  elements, but this currently is not recommended since the
+//  insert-annotations-to-source tool, which adds annotations from .jaif files
+//  into source code, adds annotations on top of existing
+//  annotations. See https://github.com/typetools/annotation-tools/issues/105 .
+//  TODO: Ensure that annotations are inserted deterministically into .jaif
+//  files. This is important otherwise developers might achieve different
+//  results (order of annotations) when running the signature inference for
+//  the same set of files.
+public class SignatureInferenceScenes {
+
+    /**
+     * Directory where .jaif files will be written to and read from.
+     * This directory is relative to where the CF's javac command is executed.
+     */
+    public final static String JAIF_FILES_PATH = "build" + File.separator +
+            "signature-inference" + File.separator;
+
+    /** Maps .jaif file paths (Strings) to Scenes. */
+    private static Map<String, AScene> scenes = new HashMap<>();
+
+    /** Maps a TypeUseLocation to the name of all annotations that should not be
```

Comment:
"to the name of all annotations" is unclear, as the value is a set. Is it fully qualified names?